mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -0,0 +1,470 @@
+               CallGraphNode *TgtNode)
+      : Call(Call), SrcNode(SrcNode), TgtNode(TgtNode) {
+    assert(SrcNode && "Expecting valid source node");
+    assert(TgtNode && "Expecting valid target node");
+  }
+
+  /// Return the func::FuncOp called by `callOp`.
+  static FunctionOpInterface getCalledFunction(const CallOpInterface &Call) {
+    if (SymbolRefAttr SymAttr = const_cast<CallOpInterface &>(Call)
+                                    .getCallableForCallee()
+                                    .dyn_cast<SymbolRefAttr>())
+      return SymbolTable::lookupNearestSymbolFrom(Call, SymAttr);
+
+    return nullptr;
+  }
+
+  CallOpInterface Call;
+  CallGraphNode *SrcNode, *TgtNode;
+};
+
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                     const ResolvedCall &RC) {
+  auto PrintCallable = [&](Operation *Op) {
+    llvm::TypeSwitch<Operation *>(Op)
+        .Case<func::FuncOp>([&](func::FuncOp Op) { OS << Op.getSymName(); })
+        .Case<gpu::GPUFuncOp>([&](gpu::GPUFuncOp Op) { Op.print(OS); })
+        .Default(
+            [&](Operation *) { assert(false && "Unhandled operation kind"); });
+  };
+
+  if (!RC.SrcNode->isExternal()) {
+    OS << "SrcNode: ";
+    PrintCallable(RC.SrcNode->getCallableRegion()->getParentOp());
+    OS << "\n";
+  }
+
+  if (!RC.TgtNode->isExternal()) {
+    OS << "TgtNode: ";
+    PrintCallable(RC.TgtNode->getCallableRegion()->getParentOp());
+    OS << "\n";
+  }
+
+  OS << "Call: ";
+  PrintCallable(ResolvedCall::getCalledFunction(RC.Call));
+  OS << "\n";
+
+  return OS;
+}
+
+/// This class represents a specific call graph SCC.
+class CallGraphSCC {
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const CallGraphSCC &);
+
+public:
+  CallGraphSCC(llvm::scc_iterator<const CallGraph *> &ParentIterator)
+      : ParentIterator(ParentIterator) {}
+
+  /// Return a range over the nodes within this SCC.
+  std::vector<CallGraphNode *>::iterator begin() { return Nodes.begin(); }
+  std::vector<CallGraphNode *>::iterator end() { return Nodes.end(); }
+
+  /// Reset the nodes of this SCC with those provided.
+  void reset(const std::vector<CallGraphNode *> &NewNodes) { Nodes = NewNodes; }
+
+  /// Remove the given node from this SCC.
+  void remove(CallGraphNode *Node) {
+    auto It = llvm::find(Nodes, Node);
+    if (It != Nodes.end()) {
+      Nodes.erase(It);
+      ParentIterator.ReplaceNode(Node, nullptr);
+    }
+  }
+
+private:
+  std::vector<CallGraphNode *> Nodes;
+  llvm::scc_iterator<const CallGraph *> &ParentIterator;
+};
+
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                     const CallGraphSCC &SCC) {
+  auto PrintCallable = [&](Operation *Op) {
+    llvm::TypeSwitch<Operation *>(Op)
+        .Case<func::FuncOp>([&](func::FuncOp Op) { OS << Op.getSymName(); })
+        .Case<gpu::GPUFuncOp>([&](gpu::GPUFuncOp Op) { Op.print(OS); })
```

Comment:
Why do these two differ?

Reply:
GPUFuncOp does not have a getSymName member function and the way to print it is by using its print member function.